### DISCUSSION TITLE
Type hints for `Security`

### DISCUSSION THREAD

**Discussion by dimaqq** [fastapi/fastapi#6071] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
def valid_token(                                                                        
    scopes: SecurityScopes,                                                             
    access_token: str = Depends(access_token_scheme),                                   
) -> Token:
    ...
    return Token(...)

@router.patch("/users/{username}")                                                                                   
async def some_endpoint(                                                          
    data: Data,                                                     
    token = Security(valid_token, scopes=["some-scope"]),                         
    username: str = Path(title="User username"),                                        
):                                                                                      
    if username == token.TYPO_HERE.username:    
        raise HTTPException(403, detail="Not allowed to change own user")
```

### Description

`mypy` is completely happy with this code, which, as far as I understand is because the return type of `valid_token` is not propagated through the `Security` wrapper.

### Wanted Solution

Some clever type hints for `Security` so that the callable return type is visible to `mypy`.

### Wanted Code

```python
pass
```

### Alternatives

The alternative is to manually type every endpoint that uses `Security`, which, as I discovered today is error prone:

```py
def valid_token(...) -> SystemToken: ...

@router.patch("/users/{username}")                                                                                   
async def some_endpoint(                                                          
    data: Data,                                                     
    token: UserToken = Security(valid_token, scopes=["some-scope"]),                         
    username: str = Path(title="User username"),                                        
):                                                                                      
    # Oops! token is actually SystemToken, not UserToken
    if username == token.username:    
        raise HTTPException(403, detail="Not allowed to change own user")
```

### Operating System

Linux

### FastAPI Version

0.89.1

### Python Version

3.11

**Comment by YuriiMotov** [comment]
It's recommended now to use the approach with `Annotated`:

```py
def valid_token(...) -> SystemToken: ...

@router.patch("/users/{username}")                                                                                   
async def some_endpoint(                                                          
    data: Data,                                                     
    token: Annotated[UserToken, Security(valid_token, scopes=["some-scope"])],
    username: Annotated[str, Path(title="User username")],
):                                                                                      
    ...
```

Generic `Security` will not help when `Annotated` approach is used.

There are a couple of PR that aimed to implement this, but thy were closed with that reason:
* https://github.com/fastapi/fastapi/pull/11255
* https://github.com/fastapi/fastapi/pull/10053
